lib/state_machines/eval_helpers.rb

```diff
@@ -71,63 +71,66 @@ module EvalHelpers
-        args = [] if (klass.method_defined?(method) || klass.private_method_defined?(method)) && object.method(method).arity == 0
-        object.send(method, *args, **, &block)
-      when Proc
+        args = [] if (klass.method_defined?(sym) || klass.private_method_defined?(sym)) && object.method(sym).arity.zero?
+        object.send(sym, *args, **, &block)
+      in Proc => proc
         args.unshift(object)
-        arity = method.arity
+        arity = proc.arity
         # Handle blocks for Procs
-        if block_given? && arity != 0
-          if [1, 2].include?(arity)
+        case [block_given?, arity]
+        in [true, arity] if arity != 0
+          case arity
+          in 1 | 2
             # Force the block to be either the only argument or the second one
             # after the object (may mean additional arguments get discarded)
             args = args[0, arity - 1] + [block]
           else
             # insert the block to the end of the args
             args << block
           end
-        elsif [0, 1].include?(arity)
+        in [_, 0 | 1]
           # These method types are only called with 0, 1, or n arguments
           args = args[0, arity]
+        else
+          # No changes needed for other cases
         end
 
         # Call the Proc with the arguments
-        method.call(*args, **)
+        proc.call(*args, **)
 
-      when Method
+      in Method => meth
         args.unshift(object)
-        arity = method.arity
+        arity = meth.arity
 
         # Methods handle blocks via &block, not as arguments
         # Only limit arguments if necessary based on arity
         args = args[0, arity] if [0, 1].include?(arity)
 
         # Call the Method with the arguments and pass the block
-        method.call(*args, **, &block)
-      when String
+        meth.call(*args, **, &block)
+      in String => str
         # Input validation for string evaluation
-        validate_eval_string(method)
+        validate_eval_string(str)
 
-        if block_given?
-          if StateMachines::Transition.pause_supported?
-            eval(method, object.instance_eval { binding }, &block)
-          else
-            # Support for JRuby and Truffle Ruby, which don't support binding blocks
-            # Need to check with @headius, if jruby 10 does now.
-            eigen = class << object; self; end
-            eigen.class_eval <<-RUBY, __FILE__, __LINE__ + 1
-                  def __temp_eval_method__(*args, &b)
-                    #{method}
-                  end
-            RUBY
-            result = object.__temp_eval_method__(*args, &block)
-            eigen.send(:remove_method, :__temp_eval_method__)
-            result
-          end
-        else
-          eval(method, object.instance_eval { binding })
+        case [block_given?, StateMachines::Transition.pause_supported?]
+        in [true, true]
+          eval(str, object.instance_eval { binding }, &block)
+        in [true, false]
```

Comment:
And (to balance out my nitpick)...

This is the sort of "gratuitous" pattern matching that I sometimes really like.  On the one hand, it loses the short-circuit evaluation of your conditions, which can be significant.  On the other hand, it changes nested if statements into something more like a "truth table" and that is sometimes a big win for readability.